pkg/controller/replicaset/replica_set.go

```diff
@@ -764,14 +767,23 @@ func (rsc *ReplicaSetController) syncReplicaSet(ctx context.Context, key string)
 	if manageReplicasErr != nil {
 		return manageReplicasErr
 	}
-	// Resync the ReplicaSet after MinReadySeconds as a last line of defense to guard against clock-skew.
+	// Plan the next availability check as a last line of defense against queue preemption (we have one queue key for checking availability of all the pods)
+	// or early sync (see https://github.com/kubernetes/kubernetes/issues/39785#issuecomment-279959133 for more info).
 	if updatedRS.Spec.MinReadySeconds > 0 &&
-		updatedRS.Status.ReadyReplicas == *(updatedRS.Spec.Replicas) &&
-		updatedRS.Status.AvailableReplicas != *(updatedRS.Spec.Replicas) {
-		nextSyncInSeconds = ptr.To(int(updatedRS.Spec.MinReadySeconds))
+		updatedRS.Status.ReadyReplicas != updatedRS.Status.AvailableReplicas {
```

Comment:
Running the check also in the middle of the rollout and not just at the end. This is useful if each pod has a varying readiness times.

Reply:
It feels like out of scope for the issue?

Reply:
I know it is a bit of an open question, and I'm likely to be missing something. However, the purpose of this change is not that clear to me in the context of the issue originally posted, so it would help to summarize the following:
- is this change relevant to fix the original issue posted
- what are the scenarios affected and not affected by this change
- does it make sense to submit a separate PR focused on this fix with dedicated tests and description

Reply:
cc @mimowo